test/distributed/test_c10d_gloo.py

```diff
@@ -702,6 +702,20 @@ def test_sparse_allreduce_basics(self):
     def test_sparse_allreduce_basics_cuda(self):
         self._test_sparse_allreduce_basics(lambda t: t.clone().cuda())
 
+    @skip_if_lt_x_gpu(2)
+    @requires_gloo()
+    def test_sparse_allreduce_cuda_dispatched(self):
+        store = c10d.FileStore(self.file_name, self.world_size)
+        dist.init_process_group(backend="gloo", store=store, rank=self.rank, world_size=self.world_size)
+        tests = simple_sparse_reduce_tests(
+            self.rank, self.world_size, num_inputs=1
+        )
+        for (inputs, outputs) in tests:
+            tensors = inputs[-1].clone().cuda()
```

Comment:
N00b question, does this mean cuda tensors can work on gloo? So we first move tensors from GPU to CPU and do the communication and move it back to CUDA?

Reply:
gloo supports `all_reduce` and `broadcast` for cuda tensors (https://pytorch.org/docs/master/distributed.html#backends). In ProcessGroupGloo implementation for all_reduce it will copy the CUDA tensors to pinned CPU tensors then performs the allreduce. So pg_gloo.all_reduce(cpu_tensor) and pg_gloo.all_reduce(cuda_tensor) are both supported.